src/node_http2_core-inl.h

```diff
@@ -630,44 +647,31 @@ inline void Nghttp2Session::RemoveStream(int32_t id) {
 
 // Implementation for Nghttp2Stream functions
 
-inline Nghttp2Stream* Nghttp2Stream::Init(
+Nghttp2Stream::Nghttp2Stream(
     int32_t id,
     Nghttp2Session* session,
     nghttp2_headers_category category,
-    int options) {
-  DEBUG_HTTP2("Nghttp2Stream %d: initializing stream\n", id);
-  Nghttp2Stream* stream = stream_free_list.pop();
-  stream->ResetState(id, session, category, options);
-  session->AddStream(stream);
-  return stream;
-}
-
+    int options) : id_(id),
+                   session_(session),
+                   current_headers_category_(category) {
+  // Allocate a fixed incoming header size based on the current local setting
+  // for max header list size.
+  uint32_t maxHeaderListSize = session->GetMaxHeaderPairs();
+  if (maxHeaderListSize == 0)
+    maxHeaderListSize = DEFAULT_MAX_HEADER_LIST_PAIRS;
+  current_headers_.AllocateSufficientStorage(maxHeaderListSize);
```

Comment:
does this mean we are preallocating an array of `maxHeaderListSize`? Do we need this to be 65535 as the default? IMHO we are wasting memory and time for each stream.

Reply:
There are two limits: the max number of header pairs, and the max number of header octets. We are preallocating an array of nhgttp2_header structs. These consist of two pointers and an 8 bit flag. Current number is a generous 1024, which can likely be tuned much much lower. The 65535 is the total number of octets and is maintained as a size_t counter only. There's no preallocation for that.

We should be able to tune the number of header pairs down to something much much smaller than 1024.

Reply:
I would go for 128

Reply:
If you’re calling `AllocateSufficientStorage()`, you kind of shouldn’t be using a `MaybeStackBuffer` in the first place – all this will give you is unused memory inside the `Nghttp2Stream` instance.

I’d suggest just going with a `std::vector` here … you don’t need to care about pre-allocating memory in that case, plus you can *still* reserve memory ahead of time using e.g. `vector.reserve(128)`.

Reply:
yeah, I keep going back and forth on this, just playing with the various options to see which is going to provide the best performance. `std::vector` is likely the best approach based on everything I've seen.